src/workers/event.rs

```diff
@@ -0,0 +1,545 @@
+        let mut current: Provider = match self.store.get(lattice_id, &id).await? {
+            Some(p) => p,
+            None => {
+                trace!("Didn't find provider in store. Creating");
+                Provider {
+                    id: provider.public_key.clone(),
+                    link_name: provider.link_name.clone(),
+                    ..Default::default()
+                }
+            }
+        };
+        debug!("Updating store with current status");
+        let status = if failed {
+            ProviderStatus::Failed
+        } else {
+            ProviderStatus::Running
+        };
+        current.hosts.insert(host_id.to_owned(), status);
+        self.store
+            .store(lattice_id, id, current)
+            .await
+            .map_err(anyhow::Error::from)
+    }
+
+    // END HANDLER FUNCTIONS
+    async fn populate_actor_info(
+        &self,
+        mut actor_ids: HashSet<String>,
+        host_id: &str,
+        count_map: &HashMap<String, usize>,
+    ) -> anyhow::Result<Vec<(String, Actor)>> {
+        let resp = self
+            .ctl_client
+            .get_claims()
+            .await
+            .map_err(|e| anyhow::anyhow!("{e:?}"))?;
+        Ok(resp
+            .claims
+            .into_iter()
+            .filter_map(|mut claim| {
+                let sub = claim.get("sub").map(|s| s.as_str()).unwrap_or_default();
+
+                actor_ids.take(sub).map(|id| {
+                    (
+                        id.clone(),
+                        Actor {
+                            id: id.clone(),
+                            name: claim.remove("name").unwrap_or_default(),
+                            capabilities: claim
+                                .remove("caps")
+                                .map(|raw| raw.split(',').map(|s| s.to_owned()).collect())
+                                .unwrap_or_default(),
+                            issuer: claim.remove("iss").unwrap_or_default(),
+                            count: [(
+                                host_id.to_owned(),
+                                count_map.get(&id).copied().unwrap_or_default(),
+                            )]
+                            .into(),
+                            ..Default::default()
+                        },
+                    )
+                })
+            })
+            .collect())
+    }
+
+    #[instrument(level = "debug", skip(self, host), fields(host_id = %host.id))]
+    async fn heartbeat_actor_update(
+        &self,
+        lattice_id: &str,
+        host: &HostHeartbeat,
+    ) -> anyhow::Result<()> {
+        debug!("Fetching current actor state");
+        let mut actors = self.store.list::<Actor>(lattice_id).await?;
+
+        let missing_actors: HashSet<String> = host
+            .actors
+            .iter()
+            .filter_map(|(id, _)| (!actors.contains_key(id)).then_some(id))
+            .cloned()
+            .collect();
+
+        // TODO: FIXME
```

Comment:
What's this? Looks like maybe there's some leftover duplicate code here?

Reply:
Yep 🤣. I discovered this yesterday when I went to add a comment

Reply:
Yep 🤣. I discovered this yesterday when I went to add a comment